hitman3.asl

```diff
@@ -1,58 +1,356 @@
-    // Starts counting along with IGT, but pauses once cutscene ends. When player loses control in game, it jumps
-    // to the value of IGT, counts again and pauses when player regains control.
-    // 48 8B 0D ? ? ? ? F3 0F 10 0D ? ? ? ? 48 81 C1 ? ? ? ? 48 8B 7C 24 ? - 3.110.1.0 sig
-    float CST: 0x03117B88, 0x898;
-    
-    // Alternate ingame time that starts after cutscene. Works for everything except start of Nightcall.
-    // Does NOT stop at end-level cutscene, only when next screen starts loading.
-    // 48 8B 1D ? ? ? ? 48 8D 4D ? 0F 29 B4 24 ? ? ? ? - 3.100.1 sig
-    // 48 8B 1D ? ? ? ? 48 8D 4D F0 - 3.110.1.0 sig
+
+}
+
+init
+{
+    Func<string, List<byte>> convertSigToByteArray = (sig) =>
+    {
+        List<byte> bytes = new List<byte>();
+
+        var stringBytes = sig.Split(' ');
+
+        foreach (var stringByte in stringBytes)
+        {
+            if (stringByte.Contains("?"))
+            {
+                bytes.Add(0xFF);
+                continue;
+            }
+
+            bytes.Add(Convert.ToByte(stringByte, 16));
+        }
+
+        return bytes;
+    };
+
+    var mainModule = modules.First();
+    var baseAddress = mainModule.BaseAddress;
+
+    var signaturesList = new[]
+    {
+        new { name = "gateExited", type = "offset", read = 0x2, sig = convertSigToByteArray("C6 85 ? ? ? ? ? 48 8B 85 ? ? ? ? 48 85 C0"), state = new int[] { 0, 0, 0 }},
+        new { name = "metadataLocation", type = "offset", read = 0x3, sig = convertSigToByteArray("48 89 9E ? ? ? ? 49 8B CD"), state = new int[] { 0, 0, 0 }},
+        new { name = "hudMissionTimeController", type = "pointer", read = 0x3, sig = convertSigToByteArray("48 8B 1D ? ? ? ? 48 85 DB 0F 84 ? ? ? ? 48 8B 43 28"), state = new int[] { 0, 0, 0 }},
+        new { name = "contractsManager", type = "pointer", read = 0x3, sig = convertSigToByteArray("48 8D 0D ? ? ? ? E8 ? ? ? ? 48 8D 4D E7 E8 ? ? ? ? F7 45 ? ? ? ? ?"), state = new int[] { 0, 0, 0 }},
+        new { name = "gameTime", type = "pointer", read = 0x3, sig = convertSigToByteArray("4C 89 0D ? ? ? ? EB 5C"), state = new int[] { 0, 0, 0 }}
+    }.ToList();
+
+    //Version check if needed in future to change signatures
+    //
+    /*	
+	var versionInfo = FileVersionInfo.GetVersionInfo(mainModule.FileName);
+	if(versionInfo.FileVersion == "3.110.1.0")
+	{
+	
+	
+	}
+	*/
+
+    var e_lfanew = memory.ReadValue<int>(baseAddress + 0x3C);
+    var SizeOfOptionalHeader = memory.ReadValue<short>(baseAddress + e_lfanew + 0x4 + 0x10);
+    var textSectionBase = baseAddress + e_lfanew + 0x18 + SizeOfOptionalHeader;
+
+    var textVirtualSize = memory.ReadValue<int>(textSectionBase + 0x08);
+    var textVirtualAddress = memory.ReadValue<int>(textSectionBase + 0x0C);
+
+    var textSectionData = memory.ReadBytes(baseAddress + textVirtualAddress, textVirtualSize);
+
+    //We probably could use internaly class SignatureScanner but it dosen't support multiple signatures at the same time
```

Comment:
As far as I understand [SignatureScanner.cs](https://github.com/LiveSplit/LiveSplit/blob/master/LiveSplit/LiveSplit.Core/ComponentUtil/SignatureScanner.cs), it _does_ allow scanning for multiple signatures, but internally it does a full scan of the read memory for each signature, right? How much memory are we processing here? Do we really get a lot of initialization speedup by implementing a "concurrent" signature matching? Just wondering if it's worth the extra complexity or if we could just use SignatureScanner without any noticeable time loss.

Reply:
Hitman is not a small executable, which is why i created my own version of the signature scanner, because in the worst case it will loop the entire executable region multiple times instead of once.

Reply:
I just did a test with StopWatch:

* the custom solution takes about 700ms to load on my PC
* the same approach with SignatureScanner takes 100ms to load

Signature scanner code (I only printed the found addresses to ensure it does find something):
```cs
    Stopwatch stopwatch = new Stopwatch();
    stopwatch.Start();

    var scanner = new SignatureScanner(game, mainModule.BaseAddress + textVirtualAddress, textVirtualSize);
    print(scanner.Scan(new SigScanTarget("C6 85 ?? ?? ?? ?? ?? 48 8B 85 ?? ?? ?? ?? 48 85 C0")).ToString());
    print(scanner.Scan(new SigScanTarget("48 89 9E ?? ?? ?? ?? 49 8B CD")).ToString());
    print(scanner.Scan(new SigScanTarget("48 8B 1D ?? ?? ?? ?? 48 85 DB 0F 84 ?? ?? ?? ?? 48 8B 43 28")).ToString());
    print(scanner.Scan(new SigScanTarget("48 8D 0D ?? ?? ?? ?? E8 ?? ?? ?? ?? 48 8D 4D E7 E8 ?? ?? ?? ?? F7 45 ?? ?? ?? ?? ??")).ToString());
    print(scanner.Scan(new SigScanTarget("4C 89 0D ?? ?? ?? ?? EB 5C")).ToString());
    print(scanner.Scan(new SigScanTarget("40 53 48 83 EC 20 48 8B D9 E8 ?? ?? ?? ?? 84 C0 74 11 48 8B CB")).ToString());
    print(scanner.Scan(new SigScanTarget("48 8D 0D ?? ?? ?? ?? 0F 11 4D E0 E8 ?? ?? ?? ?? 89 7D E8 48 8D 05 ?? ?? ?? ?? 48 89 45 E0 48 8D 55 E0")).ToString());

    stopwatch.Stop();
    print("mem load done");
    print(stopwatch.ElapsedMilliseconds.ToString());
```

Reply:
I was wrong, SignatureScanner supports multiple signatures, already changed that, thanks